### PR TITLE
A python module for contracts based verification

### LINKED ISSUE #2235: Support contracts based argument checking
**Is your feature request related to a problem? Please describe.**
Recently, we are making [a great effort in improving diagnostics and type checking](https://github.com/sql-machine-learning/sqlflow/labels/diagnostics) to enhance the experience and development efficiency of SQLFlow users. 

**Describe the solution you'd like**
Contract programming is a great idea that encourages defining formal, precise, and verifiable interface specifications for software components (function or a callable object, for example). 
> ... which extends the ordinary definition of abstract data types with preconditions, preconditions, and invariants. These specifications are referred to as "contracts", in accordance with a conceptual metaphor with the conditions and obligations of business contracts.

--- From https://en.wikipedia.org/wiki/Design_by_contract

The term 'contract programming' was coined by Bertrand Meyer in connection with his design of the Eiffel programming language. The idea is widely accepted by many other programming languages with either native support (`D`, `Scala` and many others) or 3rd-party support (`C++`  via `Boost.Contract`, `Python` via `PyContracts`, `Go`, `Java` and almost all popular languages).

In fact, contract-based verification is exactly what `SQLFlow` is doing in the [attribute](https://github.com/sql-machine-learning/sqlflow/tree/develop/pkg/sql/codegen/attribute) pkg. 

Checking the contract requirements for python functions or callables in golang is ugly, unnatural, boring, and sometimes impossible.

We propose developing a python-native contracts solution that is comprehensive,  precise, concise, and elegant.

The solution should:
1. Support checking a user-defined model via:
    - docstring
    - or decorator
    - or both
1. Support checking existing models like `XGBoost` or canned `Estimators`
1. Give precise and readable diagnostic messages to users that violate the contract requirements

Later on this day, I'll push a concise implementation for demonstration and further discussion.

### PR DESCRIPTION
**Pull request by shendiaomo**
Partially fix #2235 
This version is not compatible with python2 now.

The PR only supports a minimal subset of the idea of the contracts that I believe is enough for our scenario, the terms `contracts` and `requirements` in this implementation are mostly interchangeable.

#### Case 1: Check Existing Callables
```python
from contracts import *
import tensorflow as tf
import xgboost as xgb
my_contracts = Contracts()
my_contracts.add_requirements(tf.estimator.DNNClassifier,
                              hidden_units=Type([Natural]),
                              n_classes=Int & GreaterEqual(2),
                              optimizer=In('RMSprop', 'Adagrad'),
                              feature_columns=Type({}))

my_contracts.add_requirements(xgb.XGBModel,
                              xgb.XGBClassifier,
                              learning_rate=Float & Between(0, 1),
                              max_depth=Natural)

model_params = {
    "hidden_units": [0, 1, 2],
    "feature_column": {},
    "n_classes": 0
}

my_contracts.check_requirements(tf.estimator.DNNClassifier, model_params)
```

The above snippet will raise an exception like:
```
Traceback (most recent call last):
    ...
    raise diagnostics
__main__.Diagnostics: argument(s) didn't meet parameter requirements

hidden_units(required) must be `list of int that's > 0`. Actual: [0, 1, 2]
n_classes(optional) must be int that's >= 2. Actual: 0
feature_columns is required but is MISSING. Should be dict
feature_column is UNEXPECTED
```
The diagnostic messages are comprehensive and precise and can be further improved to be more readable.

 #### Case 2: Use Decorator to Check New Functions.
```python
@require(hidden_units=Type([Natural]), n_classes=Greater(1))
def MyDNNClassifier(...):
    # ...
```

 #### Case 3:  Requirements in Docstrings
This is not implemented now and is an important TODO.

```python
@contract
def MyDNNClassifier(...):
    ‘’‘
    :param: hidden_units
    :type: Type([Natural])
    :param: n_classes
    :type: Greater(1)
    ’‘’
    #  ...
```
The syntax detail should be carefully designed.

###### operator overloading
The design overloaded two operators (`|`, `&`) to combine different requirements. This section will use `hidden_units` of `DNNClassifier` as an example:
```python
# hidden_units must be a list of positive integers
hidden_units=IsType(list[IsInt & IsPositive])
```
##### Other Options
A pitfall of this method is that its readability is not great at a first glance. A user has to get familiar with the framework to be productive.
###### lambda
To use lambda to construct the precondition, we have to specify the error messages manually.
```python
hidden_units=require(lambda v: all(map(lambda i:  type(i) is int and i > 0, v)), "hidden_units must be a list of positive integers")
```
It's harder to write and reuse but a bit easier to generate readable diagnostic messages.

###### explicit function calls rather than overloaded operators
```python
hidden_units=IsType(list[And(IsInt,  IsPositive)])
```
Obviously, it keeps the reusability but doesn't improve the readability.

### CODE REVIEW

File: python/contracts.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+
+# Shortcuts for requirements that're often used by ML models
+Int = Type(1)
+Float = Type(1.)
+Positive = Greater(0)
+Natural = Int & Positive
+
+
+class Diagnostics(AttributeError):
```
Comment (Yancey0623) [inline python/contracts.py:212]: It seems `Diagnostics` here is only used for checking attribute error, `DiagnosticsAttributeError(Exception)` is better? And I believe we should diagnostic feature column inconsistent error, DB connection error, etc.

Comment (shendiaomo) [reply python/contracts.py:212]: > It seems `Diagnostics` here is only used for checking attribute error, `DiagnosticsAttributeError(Exception)` is better? And I believe we should diagnostic feature column inconsistent error, DB connection error, etc.

Indeed. It seems we should have a `dianostics` base and several implementations.
Another choice is to `raise` a standard `AttributeError` here and let's the `sqlflow.Diagnostics` to  catch and process it.

**Comment by typhoonzero** [comment]
If we name it "Contracts", it should follow the specification: https://www.python.org/dev/peps/pep-0316/ or we may need to change a name.

File: python/contracts.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+
+class In(Requirement):
+    """
+    In: a `Requirement` to the value of an argument. Synonym of the keyword `in`.
+    """
+    def __init__(self, *args):
+        super().__init__(lambda v: v in args,
+                         "IN ({})".format(",".join(str(i) for i in args)))
+
```
Comment (lhw362950217) [inline python/contracts.py:202]: Maybe add `NotIn` ? Likely, `NotEmpty` for **string** and container types. Of course we can add them when required.

**Comment by shendiaomo** [comment]
> If we name it "Contracts", it should follow the specification: https://www.python.org/dev/peps/pep-0316/ or we may need to change a name.

https://www.python.org/dev/peps/pep-0316/ is proposed in the early 2000s when Python itself is at an early stage(no `decorator`s and `with` e. g.) It doesn't have much practical value from the 2020s perspective. 

On the contrast, the original `contracts` syntax in Eiffel is accepted by most modern implementations.
For examples:

`Eiffel`:
```eiffel
put (x: ELEMENT; key: STRING) is
		-- Insert x so that it will be retrievable through key.
	require
		count <= capacity
		not key.empty
	do
		... Some insertion algorithm ...
	ensure
		has (x)
		item (key) = x 
		count = old count + 1
	end
```
`C++`:
```c++
template<typename T>
concept Addable = requires (T x) { x + x; }; // requires-expression

# The following is in progress
void train(const std::vector<Instance>& batch, float learning_rate)
[[expects batch.size() > 0 && learning_rate > 0 && learning_rate < 1]]
```

`D`: 
```D
int fun(ref int a, int b)
    in (a > 0)
    in (b >= 0, "b cannot be negative!")
    out (r; r > 0, "return must be positive")
    out (; a != 0)
{
    // function body
}
```
In `Eiffel`, the `require` clause introduces an input condition, or `precondition`; the `ensure` clause introduces an output condition, or `postcondition`. In `D`, the `in` and `out` work as `require` and `ensure` respectively.

SQLFlow only needs `precondition`s in the foreseeable future, so we only need to support `require` at the moment, and that's why I use the term`requirements` in most of the places.
Although the design is a simplified implementation, it still follows the good practices and principles of contract programming, thus it could be extended to full-featured implementation when necessary. 

In `C++` concepts (which can be considered as compile-time contracts), the `requires` clause introduces an interface requirement. As a result, `C++` contracts (which is under consideration) use the attribute `[[expect]]` instead of `require`. The `C++` contracts proposal is also a simplified version (no invariants), but it's still named as `contracts`

In a nutshell, `contracts` is the only appropriate name for this design.

File: python/contracts.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
+# coding: utf8
```
Comment (Yancey0623) [inline python/contracts.py:1]: I believe using DSL to express the attribute contracts can deal with the very complex scene, but users may write a very complex DSL such as `(Int & GreaterEqual(2))| (Float & Greater(3.0) & Less(10.0)) `  and maybe it's very hard to give a clear diagnostic message.

So I issued a simple way as the following prototype:

``` python
class IntList(object):
    def __init__(self):
        self.__name__ = 'Int of list'

    def __instancecheck__(cls, obj):
        return all(isinstance(x, int) for x in obj)

def diag_attribute(name, v, typ):
    if not isinstance(v, typ):
        raise TypeError("argument {0} received unexpected {1}, should be {2}".format(name, str(v), typ.__name__))


diag_attribute("n_classes", 3, int)
diag_attribute("hidden_units", [1,2, "x"], IntList())
```

which output:

``` text
Traceback (most recent call last):
  File "main.py", line 15, in <module>
    diag_attribute("hidden_units", [1,2, "x"], IntList)
  File "main.py", line 11, in diag_attribute
    raise TypeError("argument {0} received unexpected {1}, should be {2}".format(name, str(v), typ.__name__))
TypeError: argument hidden_units received unexpected [1, 2, 'x'], should be Int of list
```

and we can return the following diagnostic message to users

```text
argument hidden_units received unexpected [1, 2, 'x'], should be Int of list.
```



From this way, we can using Pythoh build-in types to be the expected type, we can also define custom type by override `__instancecheck__`,  for each custom type, we can give one clear diagnostic message like:

```text
argument A received unexpected B, should be C.
```

Comment (shendiaomo) [reply python/contracts.py:1]: I've modified the diagnostic message to be more natural, please see the (modified) original post.

**Comment by Yancey0623** [comment]
After talked with @shendiaomo ,  record some cases and requirements.

## Sample cases of diagnostic attribute error message
1. check the numerical type, and we want the following error message : 

    ``` text
    parameter [n_classes] received unexpected value `0`, should be int, greater than 1. 
    ```

1. check enum-type, and we want the following error message:

    ``` text
    parameter `my_arg` received unexpected value `x`, should be in ["a", "b", "c"]
    ``` 

1. multiple list type, we want the following error message:

    ``` text
    parameter `hidden_units` received unexpected value `3.14`, should be float of list, length greater than 2, each element should greater than 0.
    ``` 

1. there would be more than one contract for one parameter, and we want the following error message:

    ``` text
    parameter [x] received unexpected value `3.14`, which can be 
    1. int value, greater 2.
    2. list of int, length greater 3, the element of the list should be greater 2.
     ```

## Requirements

We can implement some APIs to diagnose the attribute error and show the diagnostic message on SQLFlow GUI to improve the user experience. The requirements are as the following:

1. Checking the exiting estimator, e.g. Tensorflow Estimator, XGBoost parameters.
2. when a developer wants to contribute models to Moel Zoo, he/she can use these APIs.
3. the attribute type maybe not native type, and we have not implemented an API to check it, we may have two options:
   1. Create a PR and implement a new API to check the complex type.
   2. Allow users to implement the custom type check contrast using DSL.

**Comment by shendiaomo** [comment]
> After talked with @shendiaomo , record some cases and requirements.
> 
> ## Sample cases of diagnostic attribute error message
> 1. check the numerical type, and we want the following error message :
>    ```
>    parameter [n_classes] received unexpected value `0`, should be int, greater than 1. 
>    ```
> 2. check enum-type, and we want the following error message:
>    ```
>    parameter `my_arg` received unexpected value `x`, should be in ["a", "b", "c"]
>    ```
> 3. multiple list type, we want the following error message:
>    ```
>    parameter `hidden_units` received unexpected value `3.14`, should be float of list, length greater than 2, each element should greater than 0.

list of float
>    ```
> 4. there would be more than one contract for one parameter, and we want the following error message:
>    ```
>    parameter [x] received unexpected value `3.14`, which can be 
>    1. int value, greater 2.
>    2. list of int, length greater 3, the element of the list should be greater 2.

This is great!  But I bet we have to check almost all errors at once, so indentions are needed.
>    ```
> 
> ## Requirements
> We can implement some APIs to diagnose the attribute error and show the diagnostic message on SQLFlow GUI to improve the user experience. The requirements are as the following:
> 
> 1. Checking the exiting estimator, e.g. Tensorflow Estimator, XGBoost parameters.
> 2. when a developer wants to contribute models to Moel Zoo, he/she can use these APIs.
> 3. the attribute type maybe not native type, and we have not implemented an API to check it, we may have two options:
>    
>    1. Create a PR and implement a new API to check the complex type.
>    2. Allow users to implement the custom type check contrast using DSL.

File: python/contracts.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+                                           **self._contracts[func])
+
+
+if __name__ == '__main__':
+    import tensorflow as tf
+    import xgboost as xgb
+    my_contracts = Contracts()
+    my_contracts.add_requirements(tf.estimator.DNNClassifier,
+                                  hidden_units=Type([PositiveInt]),
+                                  n_classes=Int & GreaterEqual(2),
```
Comment (Yancey0623) [inline python/contracts.py:350]: Do we actually need `AND`?
From the prototype, we can express `0<x<10` by three ways:
``` python
Int & Greater(0) & Less(10)
Int & PositiveInt & Less(10)
Int & Between(0, 10)
```
Is that what we want? The above contrast may output different diagnostic messages. And maybe we can use API instead of overload operators, that can be more clear.

``` python
x=Int(left=0, right=10, is_left_closed=False, is_right_closed=False)
```

So, I think we don't need `AND`, just `OR`.

Comment (shendiaomo) [reply python/contracts.py:350]: We need `AND`, or we have to define something like:
```python
x = Int(left=0, right=10, is_left_closed=False, is_right_closed=False)
y = Float(left=0.0, right=10.0, is_left_closed=False, is_right_closed=False)
```
Tha't not pythonic because python support `generic programming` , we don't need to implement two functions/classes with the same implementation.
So we have to define
```python
x = Number([int], left=0, right=10, is_left_closed=False, is_right_closed=False)
y = Number([int, float], left=0, right=10, is_left_closed=False, is_right_closed=False)
```
This is tediou and error-prone, and the usecase of `y` will be very common in `XGBoost`. Moreover, the implementation of `Number` has to distinguish many cases to generate a friendly error message.This is also anti-pythonic and error-prone:
```python
def Number(v, typ, left, right, is_left_closed, is_right_closed):
    assert not (left = None and right == None),   "the range is empty"
    assert typ in (int, float)
    assert something else...

    if left = None:
        if is_right_closed and v > right
            raise Diagnostics(f"should be {typ.__name__} that <= {}")
        elif is_right_closed and v >= right: 
            raise Diagnostics(f"should be {typ.__name__} that < {}")
    if right = None:
        # do something similar
    if is_right_closed and not is_left_closed:
    # ...
```
Now that we're writing python code, we should write them in a `pythonic` way, not in the `golang`'s way. The [Google Python Style](http://google.github.io/styleguide/pyguide.html) doesn't forbid operator overloading (Actually it doesn't say a word about operator overloading).  If we refer to the [Google C++ Style](https://google.github.io/styleguide/cppguide.html), it says: 
- Define overloaded operators only if their meaning is obvious, unsurprising, and consistent with the corresponding built-in operators. For example, **use | as a bitwise- or logical-or**, not as a shell-style pipe.
- **Don't go out of your way to avoid defining operator overloads.** For example, prefer to define `==`, `=`, and `<<`, rather than `Equals()`, `CopyFrom()`, and `PrintTo()`.

**Review by Yancey0623 — APPROVED** [review]
LGTM, will add some common attribute check contract base on this PR.

**Review by typhoonzero — CHANGES_REQUESTED** [review]
I still think we need further discussions, we should not require model developers to write code to verify arguments. Custom models must be Keras native.
